File: src/main/resources/eclairjs/sql/streaming/StreamingQuery.js
/*                                                                         
* Copyright 2016 IBM Corp.                                                 
*                                                                          
* Licensed under the Apache License, Version 2.0 (the "License");          
* you may not use this file except in compliance with the License.         
* You may obtain a copy of the License at                                  
*                                                                          
*      http://www.apache.org/licenses/LICENSE-2.0                          
*                                                                          
* Unless required by applicable law or agreed to in writing, software      
* distributed under the License is distributed on an "AS IS" BASIS,        
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. 
* See the License for the specific language governing permissions and      
* limitations under the License.                                           
*/ 

(function () {

    //var JavaWrapper = require(EclairJS_Globals.NAMESPACE + '/JavaWrapper');
    //var Logger = require(EclairJS_Globals.NAMESPACE + '/Logger');
    //var Utils = require(EclairJS_Globals.NAMESPACE + '/Utils');
    //var logger = Logger.getLogger("sql.streaming.StreamingQuery_js");


    
    
    /**
     * @classdesc
     * :: Experimental ::
     * A handle to a query that is executing continuously in the background as new data arrives.
     * All these methods are thread-safe.
     * @since EclairJS 0.5 Spark  2.0.0
     * @class
     * @memberof module:eclairjs/sql/streaming
     */

    var StreamingQuery = Java.type('org.eclairjs.nashorn.wrap.sql.streaming.StreamingQuery');
    //var StreamingQuery = function() {
    //	 throw "Can't instantiate abstract class - StreamingQuery";
    //\\ 	 this.logger = Logger.getLogger("StreamingQuery_js");
    //\\ 	 JavaWrapper.call(this, jvmObject);
    //
    //};
    //
    //StreamingQuery.prototype = Object.create(JavaWrapper.prototype);
    //
    //StreamingQuery.prototype.constructor = StreamingQuery;
    
    
    
    /**
     * Returns the name of the query. This name is unique across all active queries. This can be
     * set in the [[org.apache.spark.sql.DataStreamWriter DataStreamWriter]] as
     * `dataframe.writeStream.queryName("query").start()`.
     * @since EclairJS 0.5 Spark  2.0.0
     * @returns {string} 
     */
    //StreamingQuery.prototype.name = function() {
    //throw "not implemented by ElairJS";
    ////   return  this.getJavaObject().name();
    //};
    
    
    /**
     * Returns the unique id of this query. This id is automatically generated and is unique across
     * all queries that have been started in the current process.
     * @since EclairJS 0.5 Spark  2.0.0
     * @returns {number} 
     */
    //StreamingQuery.prototype.id = function() {
    //throw "not implemented by ElairJS";
    ////   return  this.getJavaObject().id();
    //};
    
    
    /**
     * Returns the {@link SparkSession} associated with `this`.
     * @since EclairJS 0.5 Spark  2.0.0
     * @returns {module:eclairjs/sql.SparkSession} 
     */
    //StreamingQuery.prototype.sparkSession = function() {
    //throw "not implemented by ElairJS";
    ////   var javaObject =  this.getJavaObject().sparkSession();
    ////   return new SparkSession(javaObject);
    //};
    
    
    /**
     * Whether the query is currently active or not
     * @since EclairJS 0.5 Spark  2.0.0
     * @returns {boolean} 
     */
    //StreamingQuery.prototype.isActive = function() {
    //throw "not implemented by ElairJS";
    ////   return  this.getJavaObject().isActive();
    //};
    
    
    /**
     * Returns the {@link StreamingQueryException} if the query was terminated by an exception.
     * @since EclairJS 0.5 Spark  2.0.0
     * @returns {module:eclairjs/sql/streaming.StreamingQueryException} 
     */
    //StreamingQuery.prototype.exception = function() {
    //throw "not implemented by ElairJS";
    ////   var javaObject =  this.getJavaObject().exception();
    ////   return new StreamingQueryException(javaObject);
    //};
    
    
    /**
     * Returns current status of all the sources.
     * @since EclairJS 0.5 Spark  2.0.0
     * @returns {SourceStatus[]} 
     */
    //StreamingQuery.prototype.sourceStatuses = function() {
    //throw "not implemented by ElairJS";
    ////   var javaObject =  this.getJavaObject().sourceStatuses();
    ////   return Utils.javaToJs(javaObject);
    //};
    
    
    /**
     *  Returns current status of the sink. 
     * @returns {module:eclairjs/sql/streaming.SinkStatus} 
     */
    //StreamingQuery.prototype.sinkStatus = function() {
    //throw "not implemented by ElairJS";
    ////   var javaObject =  this.getJavaObject().sinkStatus();
    ////   return new SinkStatus(javaObject);
    //};
    
    
    /**
     * Waits for the termination of `this` query, either by `query.stop()` or by an exception.
     * If the query has terminated with an exception, then the exception will be thrown.
     * Otherwise, it returns whether the query has terminated or not within the `timeoutMs`
     * milliseconds.
     *
     * If the query has terminated, then all subsequent calls to this method will either return
     * `true` immediately (if the query was terminated by `stop()`), or throw the exception
     * immediately (if the query has terminated with exception).
     *
     * @throws StreamingQueryException, if `this` query has terminated with an exception
     *
     * @since EclairJS 0.5 Spark  2.0.0
     * @param {number} [timeoutMs]
     * @returns {boolean} 
     */
    //StreamingQuery.prototype.awaitTermination = function(timeoutMs) {
    //throw "not implemented by ElairJS";
    ////
    ////   if (arguments[0]) {
    ////   return  this.getJavaObject().awaitTermination(timeoutMs);
    ////   } else {
    ////   return  this.getJavaObject().awaitTermination();
    ////   }
    //};
    
    
    /**
     * Blocks until all available data in the source has been processed and committed to the sink.
     * This method is intended for testing. Note that in the case of continually arriving data, this
     * method may block forever. Additionally, this method is only guaranteed to block until data that
     * has been synchronously appended data to a {@link Source}
     * prior to invocation. (i.e. `getOffset` must immediately reflect the addition).
     * @since EclairJS 0.5 Spark  2.0.0
     */
    //StreamingQuery.prototype.processAllAvailable = function() {
    //throw "not implemented by ElairJS";
    ////    this.getJavaObject().processAllAvailable();
    //};
    
    
    /**
     * Stops the execution of this query if it is running. This method blocks until the threads
     * performing execution has stopped.
     * @since EclairJS 0.5 Spark  2.0.0
     */
    //StreamingQuery.prototype.stop = function() {
    //throw "not implemented by ElairJS";
    ////    this.getJavaObject().stop();
    //};
    
    
    /**
     * Prints the physical plan to the console for debugging purposes.
     *
     * @param {boolean} [extended]  whether to do extended explain or not
     * @since EclairJS 0.5 Spark  2.0.0
     */
    //StreamingQuery.prototype.explain = function(extended) {
    //throw "not implemented by ElairJS";
    ////
    ////   if (arguments[0]) {
    ////    this.getJavaObject().explain(extended);
    ////   } else {
    ////    this.getJavaObject().explain();
    ////   }
    //};
    
    module.exports = StreamingQuery;
})();

File: server/src/main/java/org/eclairjs/nashorn/wrap/sql/streaming/StreamingQuery.java
package org.eclairjs.nashorn.wrap.sql.streaming;
/*                                                                         
* Copyright 2016 IBM Corp.                                                 
*                                                                          
* Licensed under the Apache License, Version 2.0 (the "License");          
* you may not use this file except in compliance with the License.         
* You may obtain a copy of the License at                                  
*                                                                          
*      http://www.apache.org/licenses/LICENSE-2.0                          
*                                                                          
* Unless required by applicable law or agreed to in writing, software      
* distributed under the License is distributed on an "AS IS" BASIS,        
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. 
* See the License for the specific language governing permissions and      
* limitations under the License.                                           
*/ 

import org.eclairjs.nashorn.Utils;
import org.eclairjs.nashorn.wrap.WrappedFunction;
import org.apache.log4j.Logger;
import org.eclairjs.nashorn.wrap.WrappedClass;


public class StreamingQuery extends WrappedClass {

 static Logger logger = Logger.getLogger(StreamingQuery.class);

    static WrappedFunction F_name = new WrappedFunction() {
        @Override
        public Object call(Object thiz, Object... args) {
            logger.debug("name");
            Object returnValue = null;
            org.apache.spark.sql.streaming.StreamingQuery _streamingQuery = (org.apache.spark.sql.streaming.StreamingQuery) ((StreamingQuery) thiz).getJavaObject();
            returnValue = _streamingQuery.name();
            return returnValue;
        }
    };

    static WrappedFunction F_id = new WrappedFunction() {
        @Override
        public Object call(Object thiz, Object... args) {
            logger.debug("id");
            Object returnValue = null;
            org.apache.spark.sql.streaming.StreamingQuery _streamingQuery = (org.apache.spark.sql.streaming.StreamingQuery) ((StreamingQuery) thiz).getJavaObject();
            returnValue = _streamingQuery.id();
            return returnValue;
        }
    };

    static WrappedFunction F_sparkSession = new WrappedFunction() {
        @Override
        public Object call(Object thiz, Object... args) {
            logger.debug("sparkSession");
            Object returnValue = null;
            org.apache.spark.sql.streaming.StreamingQuery _streamingQuery = (org.apache.spark.sql.streaming.StreamingQuery) ((StreamingQuery) thiz).getJavaObject();
            returnValue = _streamingQuery.sparkSession();
             return Utils.javaToJs(returnValue);
            //return new org.eclairjs.nashorn.wrap.sql.SparkSession((org.apache.spark.sql.SparkSession)returnValue);
        }
    };

    static WrappedFunction F_isActive = new WrappedFunction() {
        @Override
        public Object call(Object thiz, Object... args) {
            logger.debug("isActive");
            Object returnValue = null;
            org.apache.spark.sql.streaming.StreamingQuery _streamingQuery = (org.apache.spark.sql.streaming.StreamingQuery) ((StreamingQuery) thiz).getJavaObject();
            returnValue = _streamingQuery.isActive();
            return returnValue;
        }
    };

    static WrappedFunction F_exception = new WrappedFunction() {
        @Override
        public Object call(Object thiz, Object... args) {
            logger.debug("exception");
            Object returnValue = null;
            org.apache.spark.sql.streaming.StreamingQuery _streamingQuery = (org.apache.spark.sql.streaming.StreamingQuery) ((StreamingQuery) thiz).getJavaObject();
            returnValue = _streamingQuery.exception();
            return Utils.javaToJs(returnValue);
        }
    };

    static WrappedFunction F_sourceStatuses = new WrappedFunction() {
        @Override
        public Object call(Object thiz, Object... args) {
            logger.debug("sourceStatuses");
            Object returnValue = null;
            org.apache.spark.sql.streaming.StreamingQuery _streamingQuery = (org.apache.spark.sql.streaming.StreamingQuery) ((StreamingQuery) thiz).getJavaObject();
            returnValue = _streamingQuery.sourceStatuses();
            return Utils.javaToJs(returnValue);
        }
    };

    static WrappedFunction F_sinkStatus = new WrappedFunction() {
        @Override
        public Object call(Object thiz, Object... args) {
            logger.debug("sinkStatus");
            Object returnValue = null;
            org.apache.spark.sql.streaming.StreamingQuery _streamingQuery = (org.apache.spark.sql.streaming.StreamingQuery) ((StreamingQuery) thiz).getJavaObject();
            returnValue = _streamingQuery.sinkStatus();
            return Utils.javaToJs(returnValue);
            //return new org.eclairjs.nashorn.wrap.sql.streaming.SinkStatus((org.apache.spark.sql.streaming.SinkStatus)returnValue);
        }
    };

    static WrappedFunction F_awaitTermination = new WrappedFunction() {
        @Override
        public Object call(Object thiz, Object... args) {
            logger.debug("awaitTermination");
            Object returnValue = null;
            org.apache.spark.sql.streaming.StreamingQuery _streamingQuery = (org.apache.spark.sql.streaming.StreamingQuery) ((StreamingQuery) thiz).getJavaObject();
            
            if (args.length==0) {
              /*returnValue = */_streamingQuery.awaitTermination();
            
            } else {
            long timeoutMs =  Utils.toLong(args[0]);
              returnValue = _streamingQuery.awaitTermination(timeoutMs);
            
            }
            return returnValue;
        }
    };

    static WrappedFunction F_processAllAvailable = new WrappedFunction() {
        @Override
        public Object call(Object thiz, Object... args) {
            logger.debug("processAllAvailable");
            Object returnValue = null;
            org.apache.spark.sql.streaming.StreamingQuery _streamingQuery = (org.apache.spark.sql.streaming.StreamingQuery) ((StreamingQuery) thiz).getJavaObject();
            _streamingQuery.processAllAvailable();
            return null;
        }
    };

    static WrappedFunction F_stop = new WrappedFunction() {
        @Override
        public Object call(Object thiz, Object... args) {
            logger.debug("stop");
            Object returnValue = null;
            org.apache.spark.sql.streaming.StreamingQuery _streamingQuery = (org.apache.spark.sql.streaming.StreamingQuery) ((StreamingQuery) thiz).getJavaObject();
            _streamingQuery.stop();
            return null;
        }
    };

    static WrappedFunction F_explain = new WrappedFunction() {
        @Override
        public Object call(Object thiz, Object... args) {
            logger.debug("explain");
            Object returnValue = null;
            org.apache.spark.sql.streaming.StreamingQuery _streamingQuery = (org.apache.spark.sql.streaming.StreamingQuery) ((StreamingQuery) thiz).getJavaObject();
            
            if (args.length==0) {
              _streamingQuery.explain();
            
            } else {
            boolean extended = (boolean) args[0];
              _streamingQuery.explain(extended);
            
            }
            return null;
        }
    };


    private org.apache.spark.sql.streaming.StreamingQuery _streamingQuery;

    public StreamingQuery(org.apache.spark.sql.streaming.StreamingQuery _streamingQuery)
    {
       logger.debug("constructor");
       this._streamingQuery = _streamingQuery;
    }

    static public String getModuleName() {
        return "sql.streaming.StreamingQuery";
    }

    public boolean checkInstance(Object other) {
        return other instanceof StreamingQuery;
    }

    public Object getJavaObject() {
        return _streamingQuery;
    }

    @Override
    public String toString() {

        return _streamingQuery.toString();
    }

    public String getClassName() {
        return "StreamingQuery";
    }

    // get the value of that named property
    @Override
    public Object getMember(String name) {
        switch (name) {
            case "name":
                return F_name;
            case "id":
                return F_id;
            case "sparkSession":
                return F_sparkSession;
            case "isActive":
                return F_isActive;
            case "exception":
                return F_exception;
            case "sourceStatuses":
                return F_sourceStatuses;
            case "sinkStatus":
                return F_sinkStatus;
            case "awaitTermination":
                return F_awaitTermination;
            case "processAllAvailable":
                return F_processAllAvailable;
            case "stop":
                return F_stop;
            case "explain":
                return F_explain;
        }
        return super.getMember(name);
    }

    @Override
    public boolean hasMember(String name) {
        switch (name) {
            case "name":
            case "id":
            case "sparkSession":
            case "isActive":
            case "exception":
            case "sourceStatuses":
            case "sinkStatus":
            case "awaitTermination":
            case "processAllAvailable":
            case "stop":
            case "explain":
                return true;
        }
        return super.hasMember(name);
    }

}
